Draft feature part blog (#67)
kuzudb/kuzu-blog: src/content/post/2024-07-31-kuzu-v-0.5.0.md

categories: ["release"]
authors: ["team"]
tags: ["cypher", "extensions"]

> draft: false

Reviewer: ```suggestion
draft: true
```

```

### Json

> This release adds Json support through extension. User can now directly scan and ingest data through a json file. It is also possible to store json natively as data type in Kùzu.

Reviewer: ```suggestion
This release adds JSON support through extension. User can now directly scan and ingest data through a json file. It is also possible to store json natively as data type in Kùzu.
```

```cypher
CREATE NODE TABLE IF NOT EXISTS UW(ID INT64, PRIMARY KEY(ID));
```

#### Drop table if exists
Similar to the above case, user can drop table only if it exists. E.g.
```cypher
DROP TABLE IF EXISTS UW;
```

#### List lambda function
We added three list lambda functions
```cypher
RETRUN list_transform([1,2,3], x->x+1);
RETURN list_filter([1,2,3], x->x>1);
RETURN list_reduce([1,2,3], (x,y)->x+y) 
``` 

#### Progress bar in CLI and Kùzu Explorers
For long running queries, user can enable progress bar through
```
CALL progress_bar=true;
```

Progress bar is currently available in CLI and explorer.

#### C-API improvements
We made several improvements over C APIs in this release.

- Replace return values with out parameters, simplifying object reuse.
- Make functions that may fail now return a kuzu_state value, streamlining error handling. 
- Add data type utility functions such as `kuzu_date_t`, `kuzu_timestamp_t`, and `kuzu_interval_t`.

## Performance improvements

### MVCC

Version 0.5.0 of Kùzu brings numerous updates and improvements to the multi-version concurrency control (MVCC) protocol.

### Hint join order

> Picking an optimal join order can be challenging when it comes to a large query graph. We introduce join order hints to by-pass optimizer and enforce a specific join order. E.g.

Reviewer: @andyfengHKU could you add a sentence after the code snippet below explaining the benefit? What performance improvements did we notice? etc.

Author: Done. Also moved this to feature section instead of performance improvement section.

to hint binary join
```
MATCH (a)-[e]->(b)
WHERE a.ID = 0
HITN a JOIN (e JOIN b)
RETURN *;
```
to hint multi-way join
```
MATCH (a:person)<-[e1:knows]-(b:person)-[e2:knows]->(c:person), (a)-[e3:knows]->(c)
HINT (((a JOIN e1) JOIN b) MULTI_JOIN e2 MULTI_JOIN e3) JOIN c
RETURN COUNT(*)
```

### INT128 compression

> We extended our compression algorithm to further compress `INT128`. This compression should works for dataset with `INT128` or `UUID` column. The `INT128` compression algorithm is highly inspired by DuckDB. 

Reviewer: ```suggestion
We extended our compression algorithm to further compress `INT128`. This compression should work for dataset with `INT128` or `UUID` column. The `INT128` compression algorithm is highly inspired by the one used by DuckDB. 
```
